src/main/scala/org/apache/spark/sql/pulsar/PulsarSource.scala

```diff
@@ -54,17 +68,24 @@ private[pulsar] class PulsarSource(
 
   private var currentTopicOffsets: Option[Map[String, MessageId]] = None
 
+
   private lazy val pulsarSchema: SchemaInfo = pulsarHelper.getPulsarSchema
 
   override def schema(): StructType = SchemaUtils.pulsarSourceSchema(pulsarSchema)
 
   override def getOffset: Option[Offset] = {
-    // Make sure initialTopicOffsets is initialized
+    throw new UnsupportedOperationException(
+      "latestOffset(Offset, ReadLimit) should be called instead of this method")
+  }
+
+  override def latestOffset(startingOffset: streaming.Offset,
+                            readLimit: ReadLimit): streaming.Offset = {
     initialTopicOffsets
-    val latest = pulsarHelper.fetchLatestOffsets()
-    currentTopicOffsets = Some(latest.topicOffsets)
-    logDebug(s"GetOffset: ${latest.topicOffsets.toSeq.map(_.toString).sorted}")
-    Some(latest.asInstanceOf[Offset])
+    val admissionLimits = AdmissionLimits(readLimit)
+    pulsarHelper.latestOffsets(startingOffset, admissionLimits.get)
```

Comment:
When there is no limit we should simply use pulsarHelper.fetchLatestOffsets() without any admission control.

Reply:
You mean when readLimit is null? Is this allowed at all?

Reply:
When the user don't set a read limit, how do you handle that case?

Reply:
I thought getDefaultReadLImit is called in that case, no?

Reply:
Why do we want to do all these stats api call when there is no limit set?
We need to return a ReadAllAvailable ReadLimit in getDefaultReadLimit when maxBytesPerTrogger is not set instead of using a max integer.
Then, in latest offset
if it is a ReadMaxByte instance, we do admission control.
Otherwise just read to latest offset.